### PR TITLE
Histogram refactor, improvements.

### PR DESCRIPTION
**Pull request by tmc**
this pulls the use of OPTS and table references out of the histogram implementations.

### CODE REVIEW

**Comment by tmc** [comment]
Updated.

**Comment by tmc** [comment]
hm this lowers coverage and shouldn't, you can hold back on review

**Comment by tmc** [comment]
@okayzed ok coverage issue addressed, PTAL.

File: src/sybil/hist.go [hunk 1]
```diff
@@ -21,17 +36,16 @@ type Histogram interface {
...
 	StdDev() float64
 
-	NewHist() Histogram
-	Combine(interface{})
+	NewHist(*IntInfo) Histogram
+	Combine(Histogram)
 }
 
-func (t *Table) NewHist(info *IntInfo) Histogram {
+func NewHist(params HistogramParameters, info *IntInfo, weighted bool) Histogram {
```
Comment (okayzed) [inline src/sybil/hist.go]: why not put weight in HistogramParameters?

File: src/sybil/hist_compat.go [hunk 2]
```diff
@@ -17,9 +17,12 @@ func (hc *HistCompat) Max() int64 {
 	return hc.BasicHist.Max
 }
 
-// NewHist creates a new historgram using the same parameters.
-func (hc *HistCompat) NewHist() Histogram {
-	return hc.table.NewHist(&hc.Info)
+func (hc *HistCompat) NewHist(info *IntInfo) Histogram {
+	i := hc.Info
+	if info != nil {
```
Comment (okayzed) [inline src/sybil/hist_compat.go]: why pass an optional param?

File: src/sybil/query_spec.go [hunk 3]
```diff
@@ -103,13 +105,13 @@ func (qs *QuerySpec) NewResult() *Result {
...
 
-func (master_result *ResultMap) Combine(results *ResultMap) {
+func (master_result *ResultMap) Combine(results *ResultMap, fullMerge bool) {
 	for k, v := range *results {
 		mval, ok := (*master_result)[k]
 		if !ok {
 			(*master_result)[k] = v
 		} else {
-			mval.Combine(v)
+			mval.Combine(v, fullMerge)
```
Comment (okayzed) [inline src/sybil/query_spec.go]: why fullMerge passed in like this. maybe another way to do this, like setting an option on master_result

File: src/sybil/stats.go [hunk 4]
```diff
@@ -19,9 +19,8 @@ import "math"
 // variance, then add all those variances together
 // SSB = sum of square between groups. Take each group's averages and calculate their
 // variance against the overall average.
-func (querySpec *QuerySpec) CalculateICC() map[string]float64 {
+func (querySpec *QuerySpec) CalculateICC(params HistogramParameters) map[string]float64 {
```
Comment (okayzed) [inline src/sybil/stats.go]: where does params get used?

Comment (tmc) [reply src/sybil/stats.go]: whoops, it's not.

File: src/sybil/hist_compat.go (hunk 2 above)
Comment (tmc) [reply src/sybil/hist_compat.go]: the idea here is that the params from hc would be used unless a non-nil value is passed in. We could just always accept them and have the callers always provide. That would look like hc.NewHist(hc.Info)

**Comment by tmc** [comment]
the bug was just in the test so it wasn't running with loghist correclty.

Addressed some issues and rebased. PTAL @okayzed

File: src/sybil/query_spec.go [hunk 5]
```diff
@@ -148,31 +157,19 @@ func (rs *Result) Combine(nextResult *Result) {
 
 	// combine histograms...
 	for k, h := range nextResult.Hists {
+		ph, ok := rs.Hists[k]
 
-		// If we are doing a node aggregation, we have a MERGE_TABLE
```
Comment (okayzed) [inline src/sybil/query_spec.go]: this comment should still remain somewhere (or something similar)

**Comment by tmc** [comment]
#69 fixes the regression in #61

**Comment by tmc** [comment]
@okayzed ok comment restored.

File: src/sybil/query_spec.go [hunk 6]
```diff
@@ -134,10 +134,14 @@ func fullMergeHist(h, ph Histogram) Histogram {
 	return nh
 }
 
-func extentsMatch(l, r Histogram) bool {
+// fastMergable indicates if two histograms can be merged on the fast path vs a full merge.
+func fastMergable(l, r Histogram) bool {
```
Comment (okayzed) [inline src/sybil/query_spec.go]: mergeable*

**Pull request merged**